docs/design/generics/details.md

```diff
@@ -2592,19 +2592,25 @@ this interface while `.Self` refers to the associated facet currently being
 defined.
 
 ```carbon
+interface Container;
+constraint SliceConstraint(E:! type, S:! Container);
+
 interface Container {
   let ElementType:! type;
   let IteratorType:! Iterator where .ElementType = ElementType;
 
-  let SliceType:! Container
-      where .ElementType = ElementType and
-            // `.Self` means `SliceType`.
-            .SliceType = .Self;
+  // `.Self` means `SliceType`.
+  let SliceType:! SliceConstraint(ElementType, .Self);
```

Comment:
Hm, how do we type-check this call to `SliceConstraint`? We don't know that `.Self impls Container` here. And
```
  let SliceType:! Container & SliceConstraint(ElementType, .Self);
```
doesn't help, because while type-checking the type expression we still have `.Self:! type`. This would work:
```suggestion
  let SliceType:! Container where .Self is SliceConstraint(ElementType, .Self);
```
and I don't think that's a bad workaround. In fact, I think you can drop the `.Self` parameter in this case and define `SliceConstraint` as 
```
constraint SliceConstraint(E:! type) {
  // Note, might need to use `.Self:! Container`, not `Self:! type` for the better type,
  // even though they would have the same value here.
  extend Container where .ElementType = E and
                         .SliceType = .Self;
}
```

Should this be a context where we have implied constraints? That is, should we pick up an implicit `.Self impls Container` constraint from the use of `.Self` as the argument to `S:! Container` here? I think that would make sense, but currently we only get implied constraints on a generic function. (I'm not sure whether we should only allow implied constraints on `.Self` or on other associated constants for the interface too.)

Example:

```carbon
interface Hashable {
  let KeyType:! type;
  // or `let ValueType:! type; require HashMap(KeyType, ValueType) impls type;`
  let ValueType:! type where HashMap(KeyType, .Self) impls type;
}
```
would be a useful shorthand for
```carbon
interface Hashable {
  let KeyType:! type;
  let ValueType:! type;
  // (whatever constraints are implied by `HashMap`)
  require KeyType impls HashMapKey;
  require ValueType impls HashMapValue;
}
```

Reply:
>  In fact, I think you can drop the `.Self` parameter in this case and define `SliceConstraint` as
> 
> ```
> constraint SliceConstraint(E:! type) {
>   // Note, might need to use `.Self:! Container`, not `Self:! type` for the better type,
>   // even though they would have the same value here.
>   extend Container where .ElementType = E and
>                          .SliceType = .Self;
> }
> ```

I realized I could have included the `.Self` inside `SliceConstraint`, but I was trying to make an example of an interface using `.Self` and `Self` to highlight how they are different.

Reply:
Filed #3256 for the question about implied constraints.